Update miniconda builds

clbarnes opened the pull request:
We're pretty behind on miniconda versions. This includes all new major.minor.patch versions since 4.3.30, as well as a script to automatically populate new miniconda versions in the future.

- Ignores releases below 4.3.30
- Ignores "-latest" releases
- Ignores sub-patch releases if that major.minor.patch already exists
  - But otherwise, takes the latest sub-patch release for given OS/arch
- Assumes all miniconda3 releases > 4.3.30 default to python 3.6 (correct at time of writing)

### Description

- miniconda[23] versions 4.3.31 through 4.6.14.

### Tests

I am not sure how best to test adding new builds; is there anything more ergonomic than just trying to build each one, one at a time?

See also #1360 for a discussion on miniconda-latest.

mariusvniekerk commented:
Anything remaining on this @yyuu before we can merge?

yyuu commented on plugins/python-build/scripts/add_miniconda.py:
```diff
@@ -0,0 +1,261 @@
+    py_version: Optional[PyVersion] = None
+
+    @classmethod
+    def from_filestem(cls, stem, md5, py_version=None):
```
Please consider using sha256sum in lieu of md5sum. The current code actually can validate MD5 checksum, however, it's just left for backward compatibility and no new code should use the legacy hashing algo.

clbarnes replied:
I used md5 because that's the checksum they have on the installer archive. If there's an equally accessible sha256 checksum I'll look into it.

yyuu replied:
I still prefer to use SHA256 in general as MD5 is not considered as safe since awhile ago. However, I can understand that it's not possible to generate SHA256 without downloading remote contents in case the official doesn't provide it. It might be nice if generated file will have some shell script comment like `# FIXME: please replace this md5sum by sha256sum` maybe...? 🤔

yyuu commented on plugins/python-build/scripts/add_miniconda.py:
```diff
@@ -0,0 +1,261 @@
+#!/usr/bin/env/python3.7
```
It might be little funny if a Python script packaged with pyenv is configured to skip using pyenv 😛

clbarnes replied:
Oops, that shouldn't have the slash before python.

I don't think this necessarily skips using pyenv though - wouldn't `/usr/bin/env python3.7` use whichever python3.7 is first in the path, including whatever pyenv shims in?

yyuu replied:
It actually depends on how `$PATH` is configured for the caller's shell, however, having `/usr/bin/env python3.7` in shebang would work sufficiently for the purpose.

clbarnes commented:
I still haven't tested most of the builds this adds - is there a good strategy for that?

Any thoughts on the sub-patch builds?

clbarnes commented:
- rebased on master
- fixed shebang of script
- re-ran script to add miniconda[23]-4.7.10

mariusvniekerk commented:
@yyuu What more is needed here?

clbarnes commented:
I need to update it to use python 3.7 for recent miniconda versions, and test all of them. Is there a good strategy for testing large numbers of builds?

clbarnes commented:
I have rebased this on master and fixed the py37 issue.

The review comment from @yyuu can't be addressed, as the usage of md5sum is anaconda's choice, not mine.

All this needs now is testing, but I can't find any information on how to test installed python versions without my own (mainline) pyenv installation interfering; see #1445 .

`bats` testing gets the same two failures on master and on this branch:

```
 ✗ sys.executable with system version (#98)                                                                                                     
   (from function `flunk' in file test/test_helper.bash, line 42,                                                                               
    from function `assert_success' in file test/test_helper.bash, line 47,                                                                      
    in test file test/exec.bats, line 90)                                                                                                       
     `assert_success "${system_python}"' failed                                                                                                 
   command failed with exit status 1                                                                                                            
 ✗ /tmp/pyenv.NZj/root/shims:/home/barnesc/code/pyenv/test/libexec:/home/barnesc/code/pyenv/test/../libexec:/tmp/pyenv.NZj/bin:/usr/bin:/bin:/us
r/sbin:/sbin:/usr/local/bin is not modified with system Python                                                                                  
   (from function `flunk' in file test/test_helper.bash, line 42,                                                                               
    from function `assert_success' in file test/test_helper.bash, line 47,                                                                      
    in test file test/exec.bats, line 102)                                                                                                      
     `assert_success "$PATH"' failed                                                                                                            
   command failed with exit status 1   
```

clbarnes commented:
Rebased on master again. Still waiting for documentation or advice on how to test new versions.

clbarnes commented:
Rebasing this on PR #1548 locally allowed me to test a handful of these very easily and in complete isolation.

yyuu commented on plugins/python-build/scripts/add_miniconda.py:
```diff
@@ -0,0 +1,265 @@
+
+if __name__ == "__main__":
+    parser = ArgumentParser()
+    parser.add_argument("-d", "--dry_run", action="store_true")
```
trivial - I would _feel_ that `--dry-run` is more intuitive than underscore name 🤔

clbarnes replied:
Yes, I dimly recall some time in the past argparse not being able to handle dashes in the middle of an option name but that doesn't seem to be the case now.

yyuu commented on plugins/python-build/scripts/add_miniconda.py:
```diff
@@ -0,0 +1,265 @@
+#!/usr/bin/env python3.7
```
It would be nice if there's some usage example of this script. In fact, just running the script without giving any arguments will let the script to update the python-build's build definitions for miniconda based on the contents of Miniconda's download page, however, it's not very obvious without looking into the script itself. Since this is just some sort of utility script for maintainers, it doesn't need to be some individual documentation file, however, it's nice if there's some practical usage example is in somewhere (e.g. show message on `--help`, etc.)

clbarnes replied:
Yep - I think at time of writing 3.8 wasn't out, and 3.7 was pretty new so I wanted to be explicit about it. Now I'm happy to just specify 3 in the shebang and document that it's 3.7+.

Usage example is a good idea. I'll write a module docstring and have that reported by the help message.

yyuu commented on plugins/python-build/scripts/add_miniconda.py:
```diff
@@ -0,0 +1,265 @@
+
+
+if __name__ == "__main__":
+    parser = ArgumentParser()
```
trivial - it's nice if argparse can take `--verbose` or `--debug` to enable debug logging.

clbarnes replied:
Sure, I can add more verbose logging.

clbarnes commented:
I think I'll add a readme with a usage example etc.

muendelezaji commented:
It appears there might be a problem in the merged code - there are duplicate architecture entries in the [miniconda2-4.7.12](https://github.com/pyenv/pyenv/pull/1361/files/81e797be12e86e70bf5e332b3ac15512d65ad950#diff-e51d5e7af8a59099804d79c11541a472) and [miniconda3-4.7.12](https://github.com/clbarnes/pyenv/blob/81e797be12e86e70bf5e332b3ac15512d65ad950/plugins/python-build/share/python-build/miniconda3-4.7.12) python-build files. This manifests as the 4.7.12 shadowing the 4.7.12.1 version i.e. unable to install 4.7.12.1.

Is that intended, or should there be a separate 4.7.12.1 python-build file?

clbarnes commented:
As I understand it, pyenv does not and has never handled distinguishing subpatch versions for the \*condas. As such, this script picks the latest available subpatch version when any given patch version is first picked up. It doesn't overwrite old versions when new subpatch versions are introduced, partially for reasons specified in #1360, partially to prevent it being destructive. Something is very wrong with anaconda's numbering scheme if anybody's workflow would be broken by having a newer subpatch version than they expected. 

I'm open to supporting subpatch versions and it would probably be trivial to update the script to do so, but this is a bigger new feature than I wanted to introduce with this PR. Other alternatives for handling subpatch versions are welcome in #1360.

muendelezaji commented:
> As such, this script picks the latest available subpatch version when any given patch version is first picked up.

Just to clarify, I also expected this behaviour on a fresh install when not overwriting a previous miniconda. But the order in the shell script means that the case statement returns the first match (i.e. 4.7.12) and never gets to the latest subpatch version.

For example, from [miniconda3-4.7.12](https://github.com/clbarnes/pyenv/blob/81e797be12e86e70bf5e332b3ac15512d65ad950/plugins/python-build/share/python-build/miniconda3-4.7.12)
```shell
# ...
# matches this first
"Linux-x86_64" )
  install_script "Miniconda3-4.7.12-Linux-x86_64" "https://repo.anaconda.com/miniconda/Miniconda3-4.7.12-Linux-x86_64.sh#0dba759b8ecfc8948f626fa18785e3d8" "miniconda" verify_py37
  ;;

# ...
# never gets here
"Linux-x86_64" )
  install_script "Miniconda3-4.7.12.1-Linux-x86_64" "https://repo.anaconda.com/miniconda/Miniconda3-4.7.12.1-Linux-x86_64.sh#81c773ff87af5cfac79ab862942ab6b3" "miniconda" verify_py37
  ;;

# ...
```

So the question was, is this deliberately written that way, should they be swapped around, or any other approach (removing subpatch completely)?

clbarnes commented:
Yes, you're right. That wasn't intentional. I'll see what I can do.
